### PR TITLE
Project registration

### LINKED ISSUE #437: Project Modal Form
### Pre-requisites:
- [x] I looked through both [open and closed issues](../issues?utf8=✓&q=is%3Aissue) and did not find another request for the same or similar feature.

## Description

- Rename the "Pre-Register" button to "Register"
- Rename "My Project" nav item to "Home"
- Create a form for when they click the Register button (inside modal)
- Fields for name, short description, and table number (not required)
- On any server errors show an error Alert component asking them to try again or come talk to us

### PR DESCRIPTION
**Pull request by dre8597**
## Pre-Requisites
- [X] Yes, I updated [Authors.md](../blob/main/Authors.md) **OR** this is not my first contribution
- [X] Yes, I included and/or modified tests to cover relevent code **OR** my change is non-technical
- [X] Yes, I wrote this code entirely myself **OR** I properly attributed these changes in [Third Party Notices](../blob/main/THIRD-PARTY-NOTICES.txt)

## Description of Changes

*NOT COMPLETE* needs to be updated with changes from #436 

Adding a registration modal to allow teams to register their projects

Screen Shot 2021-11-11 at 5 53 04 PM

Screen Shot 2021-11-11 at 5 54 00 PM

Screen Shot 2021-11-11 at 5 54 20 PM

## Related Issues

Closes #437

### CODE REVIEW

File: packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+      try {
+        if (!res.ok) {
+          throw new Error();
+        }
+
+        const data = await res.json();
+        setProject(data);
+      } catch (err) {
+        // TODO print/log error message to show we were unable to pull their project data
```
Comment (SpencerKaiser) [inline packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx:42]: Just show an `Alert` to the user with the error message the backend sends us

File: packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            isInvalid={!!formik.errors.name}
+            onChange={formik.handleChange}
+            onBlur={formik.handleBlur}
+            disabled={formik.isSubmitting}
+          />
+          <FormHelperText color="red.500">{formik.errors.name}&nbsp;</FormHelperText>
+        </FormControl>
+        <FormControl id="description">
+          <FormLabel>Description</FormLabel>
+          <Input
```
Comment (SpencerKaiser) [inline packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx:105]: See if a `TextArea` will work here - if not, let's chat and see if it's worth the effort

File: packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            type="text"
+            placeholder="My project will..."
```
Comment (SpencerKaiser) [inline packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx:107]:
```suggestion
            placeholder="With the new American Airlines Mac app, you get the information you need exactly when you need it. Curious about traffic to the airport? Wondering if a better seat is available? All this and more is at your fingertips."
```

File: packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        body: JSON.stringify({
+          inputConfig: values,
+        }),
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      });
+      if (!res.ok) {
+        setServerError('Something went wrong, please try again later...');
+      }
```
Comment (SpencerKaiser) [inline packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx:77]: On success, we probably either want to show a success modal or we want to invoke a callback that this component received as a prop

@johnkahn any preference?

File: packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            onBlur={formik.handleBlur}
+            disabled={formik.isSubmitting}
+          />
+          <FormHelperText color="red.500">{formik.errors.description}&nbsp;</FormHelperText>
+        </FormControl>
+        <FormControl id="tableNumber">
+          <FormLabel>Table Number</FormLabel>
+          <Input
+            type="text"
+            placeholder="Enter value..."
```
Comment (SpencerKaiser) [inline packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx:120]:
```suggestion
            placeholder="42"
```

File: packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  }, [formik.submitCount, validateWhileTyping]);
+
+  return (
+    <form onSubmit={formik.handleSubmit}>
+      <VStack alignItems="stretch">
+        <FormControl id="name">
+          <FormLabel>Name</FormLabel>
+          <Input
+            type="text"
+            placeholder="Your project name"
```
Comment (SpencerKaiser) [inline packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx:94]:
```suggestion
            placeholder="American Mobile App"
```

File: packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  tableNumber?: number;
+}
+
+export const RegistrationForm: React.FC = () => {
+  const [serverError, setServerError] = useState('');
+  const [project, setProject] = useState<ProjectProps>({});
+  const [validateWhileTyping, setValidateWhileTyping] = React.useState(false);
+
+  useEffect(() => {
+    const fetchProjects = async () => {
```
Comment (SpencerKaiser) [inline packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx:31]:
```suggestion
    const fetchProject = async () => {
```

File: packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+      }
+    };
+
+    void fetchProjects();
+  }, []);
+
+  const RegistrationSchema = Yup.object().shape({
+    name: Yup.string().required('Name is required'),
+    description: Yup.string(),
+    tableNumber: Yup.number().typeError('table number need to be digi'),
```
Comment (SpencerKaiser) [inline packages/web/src/components/ProjectRegistrationModal/RegistrationForm.tsx:54]:
```suggestion
    tableNumber: Yup.number().typeError('Table number must be an integer'),
```

**Review by johnkahn — APPROVED** [review]

**Pull request merged**